File: apps/web/src/types/connectors.ts
import type { SourcePlatform, NormalizedSecurityEvent, EventCategory } from './events'

/**
 * Connector / ingestion foundation types (v1.3.0).
 *
 * Defines the boundary for how future security data enters SentinelIQ. Only the mock
 * connector is implemented; Sentinel/Splunk/Elastic/EDR/IdP are declared placeholders that
 * report `not_configured` and never call a real API.
 */

export type ConnectorStatus = 'connected' | 'not_configured' | 'error' | 'disabled'

/** How an ingestion run executed (never real yet). */
export type ConnectorMode = 'mock' | 'dry_run' | 'real_placeholder'

export type ConnectorCapability =
  | 'authentication'
  | 'process'
  | 'network'
  | 'identity'
  | 'cloud'
  | 'alerts'
  | 'rules'
  | 'query'

export interface ConnectorSource {
  name: string
  kind: 'table' | 'index' | 'stream'
  category?: EventCategory
}

/** Static + runtime metadata describing a connector (what the Data Sources page renders). */
export interface ConnectorMetadata {
  id: string
  name: string
  platform: SourcePlatform
  description: string
  status: ConnectorStatus
  mode: ConnectorMode
  capabilities: ConnectorCapability[]
  recordsAvailable: number
  lastSync: string | null
  /** Plain-language note (e.g. "Planned — real credentials are future work"). */
  note?: string
}

export interface TestConnectionResult {
  ok: boolean
  status: ConnectorStatus
  message: string
}

/**
 * The provider interface every connector implements. Methods are synchronous for the
 * deterministic mock; a future real connector would return Promises (the UI already treats
 * results as data, so widening to async later is a localized change).
 */
export interface SecurityDataConnector {
  meta: ConnectorMetadata
  listSources(): ConnectorSource[]
  testConnection(): TestConnectionResult
  fetchSampleEvents(limit?: number): NormalizedSecurityEvent[]
  fetchAlerts(): NormalizedSecurityEvent[]
  runQuery(renderedQuery: string): NormalizedSecurityEvent[]
  normalize(raw: Record<string, unknown>): NormalizedSecurityEvent
  getLastSync(): string | null
}

// ── Ingestion runs ────────────────────────────────────────────────────────

export type IngestionStatus = 'pending' | 'running' | 'success' | 'failed'

export interface IngestionRun {
  id: string
  connectorId: string
  connectorName: string
  startedAt: string
  completedAt: string | null
  status: IngestionStatus
  recordsFetched: number
  recordsNormalized: number
  alertsCreated: number
  errors: string[]
  mode: ConnectorMode
}

File: apps/api/src/schemas/connectors.py
"""
Connector / ingestion schemas (v1.3.0).

Deterministic mock connector metadata + ingestion-run models. No real credentials, no real
API calls — real platforms are declared placeholders that report `not_configured`.
"""
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel


class ConnectorMeta(BaseModel):
    id: str
    name: str
    platform: str
    description: str
    status: str           # connected | not_configured | error | disabled
    mode: str             # mock | dry_run | real_placeholder
    capabilities: list[str]
    recordsAvailable: int
    lastSync: Optional[str] = None
    note: Optional[str] = None


class TestConnectionResult(BaseModel):
    ok: bool
    status: str
    message: str


class IngestionRun(BaseModel):
    id: str
    connectorId: str
    connectorName: str
    startedAt: str
    completedAt: Optional[str] = None
    status: str           # pending | running | success | failed
    recordsFetched: int = 0
    recordsNormalized: int = 0
    alertsCreated: int = 0
    errors: list[str] = []
    mode: str = "mock"


class SyncRequest(BaseModel):
    """Optional run payload the frontend can post; if absent the backend generates one."""
    run: Optional[dict[str, Any]] = None


class NormalizedEvent(BaseModel):
    """Vendor-neutral normalized event (mirrors the frontend NormalizedSecurityEvent)."""
    id: str
    timestamp: str
    sourcePlatform: str
    sourceProduct: str
    sourceType: str
    sourceTableOrIndex: str
    eventCategory: str
    eventName: str
    severity: Optional[str] = None
    user: Optional[str] = None
    host: Optional[str] = None
    ip: Optional[str] = None
    process: Optional[str] = None
    commandLine: Optional[str] = None
    country: Optional[str] = None
    ruleName: Optional[str] = None
    tactic: Optional[str] = None
    technique: Optional[str] = None
    linkedAlertIds: list[str] = []
    linkedInvestigationIds: list[str] = []
